nixos/modules/config/networking.nix

```diff
@@ -188,11 +214,23 @@ in
 
         # /etc/hosts: Hostname-to-IP mappings.
         "hosts".text =
+          let oneToString = set : ip : ip + " " + concatStringsSep " " ( getAttr ip set );
+              allToString = set : concatMapStringsSep "\n" ( oneToString set ) ( attrNames set );
+              userLocalHosts = optionalString
+                ( builtins.hasAttr "127.0.0.1" cfg.hosts )
+                ( concatStringsSep " " ( remove "localhost" cfg.hosts."127.0.0.1" ));
+              userLocalHosts6 = optionalString
+                ( builtins.hasAttr "::1" cfg.hosts )
+                ( concatStringsSep " " ( remove "localhost" cfg.hosts."::1" ));
+              otherHosts = allToString ( removeAttrs cfg.hosts [ "127.0.0.1" "::1" ]);
+              maybeFQDN = optionalString ( cfg.fqdn != null ) cfg.fqdn;
```

Comment:
Wouldn't it make more sense to put the following into `config` instead?

```nix
{
  networking.hosts = {
    "127.0.0.1" = [ "localhost" ];
  } // optionalAttrs cfg.enableIPv6 {
    "::1" = [ "localhost" ];
  };
}
```

That way the value in `environment.etc.hosts.text` can be reduced to:

```nix
let
  mkEntry = ip: hosts: "${ip} ${concatMapStringsSep " " hosts}";
in ''
  ${concatStringsSep "\n" (mapAttrsToList mkEntry cfg.hosts)}
  ${cfg.extraHosts}
''
```